File: ml-benchmark/synduce/ptree/maxheads.f
Inductive Tree = telt Int | tnode {Int, Tree, Tree};
Inductive PTree = pelt Int | pnode {Int, List}
    with List = elt PTree | cons {PTree, List};

max = \a: Int. \b: Int. if < a b then b else a;

repr = fix (
  \f: Tree -> Compress PTree. \x: Tree.
  match x with
    telt a -> pelt a
  | tnode {a, l, r} -> 
    let repr_list = fix (
        \g: Tree -> Compress List. \y: Tree.
        match y with
          telt a -> let z = pelt a in elt z
        | tnode {a, l, r} -> cons {pelt a, cons {f l, g r}}
        end
    ) in
    pnode {a, cons {f l, repr_list r}}
  end
);

spec = fix (
  \f: PTree -> Int. \x: PTree.
  match x with
    pelt a -> a
  | pnode {a, l} -> 
    let maxh_aux = fix (
        \g: List -> Int. \y: List.
        match y with
          elt a -> f a
        | cons {hd, tl} -> f hd
        end
    ) in 
    max a (maxh_aux l)
  end
);

main = \x: Tree. spec (repr x);

File: ml-benchmark/synduce/ptree/maxheads.ml
type tree = Telt of int | Tnode of int * tree * tree

type 'a ptree = Pelt of int | Pnode of int * 'a
type plist = Elt of plist ptree | Cons of plist ptree * plist

let max a b = if a < b then b else a

val repr: tree -> (plist ptree) compress
let rec repr x =
  match x with
  | Telt a -> Pelt a
  | Tnode (a, l, r) ->
      let rec repr_list y =
        match y with
        | Telt a ->
            let z = Pelt a in
            Elt z
        | Tnode (a2, l2, r2) ->
            Cons (Pelt a2, Cons (repr l2, repr_list r2))
      in
      Pnode (a, Cons (repr l, repr_list r))

let rec spec x =
  match x with
  | Pelt a -> a
  | Pnode (a, l) ->
      let rec maxh_aux y =
        match y with
        | Elt a2 -> spec a2
        | Cons (hd, tl) -> spec hd
      in
      max a (maxh_aux l)

let program x = spec (repr x)